python/tvm/autotvm/task/task.py

```diff
@@ -0,0 +1,340 @@
+"""Definition of task function.
+
+Task can be constructed from tuple of func, args, and kwargs.
+func is a state-less function, or a string that
+registers the standard task.
+"""
+
+from functools import wraps
+
+import numpy as np
+
+from ... import build, nd, tensor, expr, target as _target
+from ..template import DispatchContext, ConfigSpace, ApplyConfig, dispatcher
+from ..util import get_const_tuple, get_const_int
+
+
+class Task(object):
+    """A Tunable Task
+
+    Parameters
+    ----------
+    name: str
+        The name of the task.
+    args: list
+        Positional argument of func
+    """
+    def __init__(self, name, func, args):
+        self.name = name
+        self.args = args
+        self.kwargs = {}  # currently unused
+
+        self.func = func
+
+        # init null config space
+        self.config_space = None
+        self.ref_input, self.ref_output = None, None
+
+        # auxiliary info, available after `init_space` is called
+        self.workload = None
+        self.flop = None
+        self.target = None
+        self.target_host = None
+
+    def init_ref_data(self, arg_bufs, check_correctness=False):
+        """Get reference input/output by building a llvm function
+
+        Parameters
+        ----------
+        arg_bufs: Array of tvm.Tensor
+            args to extract type and shape for input
+        check_correctness: bool
+            whether check correctness by a llvm reference implementation
+        """
+        self.ref_input = [np.random.uniform(size=get_const_tuple(x.shape)).astype(x.dtype)
+                          for x in arg_bufs]
+        if not check_correctness:
+            return
+        with _target.create("llvm"):
+            s, args = self.instantiate(self.config_space.get(0))
+        func = build(s, args, "llvm")
+        tvm_buf = [nd.array(x) for x in self.ref_input]
+        func(*tvm_buf)
+        self.ref_output = [x.asnumpy() for x in tvm_buf]
+
+    def init_space(self, target, target_host=None, template_key=None):
```

Comment:
We might want to make task immutable, in that case, init_space of a task should return a new Task, with these specializations